### PR TITLE
Hard Kutta TE Constraint: physics-enforced pressure continuity at trailing edges

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model predicts surface pressure at each node independently, with no explicit constraint that upper and lower surface pressures must match at the trailing edge — the **Kutta condition**, a fundamental physics requirement for lifting surfaces with sharp trailing edges.

Trailing edge pressure errors cascade along the pressure recovery region and are disproportionately represented in p_tan errors (tandem inter-foil wake region experiences TE-driven pressure mismatches from both foils). Enforcing the Kutta condition as a hard differentiable post-processing step + a soft TE-region smoothness loss should directly reduce p_tan by penalizing discontinuities the backbone is currently free to produce.

**Physics basis:** For a sharp trailing edge, the Bernoulli equation requires pressure continuity (Cp_upper = Cp_lower at TE). Violating this implies infinite velocity at the TE, which is unphysical. Current model has no mechanism to enforce this.

**Expected impact:** p_tan -3–8%, p_in -1–3%. Zero new learned parameters — pure physics projection + loss term.

## Instructions

The `compute_te_features` function (line 256 in `train.py`) already identifies the trailing edge node per foil via argmax of x-coord among surface nodes. This gives `fore_te_idx`, `aft_te_idx` (currently computed internally but not returned).

### Step 1: Expose TE node indices

Modify `compute_te_features` to also return `fore_te_idx` and `aft_te_idx`:
```python
# Change return from:
return torch.stack([...], dim=-1), fore_te_x, fore_te_y
# To:
return torch.stack([...], dim=-1), fore_te_x, fore_te_y, fore_te_idx, aft_te_idx
```
Update all 4 call sites accordingly.

### Step 2: Find TE-region surface nodes

For each batch element, find the **K=8 nearest surface nodes to each TE node** (covering both upper and lower TE surfaces, ~4 nodes per side):

```python
# raw_xy: [B, N, 2], is_surface: [B, N] bool
fore_te_xy = raw_xy[torch.arange(B), fore_te_idx]  # [B, 2]
dist_to_fore_te = ((raw_xy - fore_te_xy[:, None, :]) ** 2).sum(-1)  # [B, N]
dist_to_fore_te[~is_surface] = 1e9  # mask non-surface nodes
fore_te_region_idx = dist_to_fore_te.topk(K, dim=1, largest=False).indices  # [B, K]
# Repeat for aft TE; zero out contribution for single-foil samples
```

### Step 3: Hard constraint (differentiable TE projection)

After the model output head, replace the K TE-region node predictions with their mean:

```python
# pred_surface: [B, N] model output predictions
fore_te_preds = pred_surface.gather(1, fore_te_region_idx)  # [B, K]
te_region_mean = fore_te_preds.mean(dim=1, keepdim=True)    # [B, 1]
pred_surface = pred_surface.scatter(1, fore_te_region_idx,
    te_region_mean.expand(-1, K))  # replaces K TE-region preds with their mean
# Repeat for aft TE (masked for single-foil)
```
Apply this BEFORE computing the main MAE loss so backprop flows through the projection.

### Step 4: Kutta smoothness loss

Penalize the spread of predictions among the K TE-region nodes (max-min):

```python
kutta_fore = (fore_te_preds.max(dim=1).values - fore_te_preds.min(dim=1).values) ** 2  # [B]
kutta_aft = ... * is_tandem   # [B]
kutta_loss = (kutta_fore + kutta_aft).mean()
total_loss = main_loss + cfg.kutta_loss_weight * kutta_loss
```

### Step 5: New config flags

```python
kutta_loss_weight: float = 0.0       # Kutta TE smoothness loss weight
kutta_hard_constraint: bool = False  # Hard TE region averaging projection
kutta_region_k: int = 8             # Nearest TE nodes to enforce continuity
```

### Step 6: Run 3 variants — `--wandb_group kutta-te-constraint`

```bash
# Variant 1: Hard + Soft (λ=0.5) — run this first with seed 42
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent edward --wandb_name "edward/kutta-hard-soft-s42" \
  --wandb_group kutta-te-constraint --seed 42 \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --kutta_hard_constraint --kutta_loss_weight 0.5 --kutta_region_k 8

# Variant 2: Soft only (λ=1.0) — run after variant 1
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent edward --wandb_name "edward/kutta-soft-only-s42" \
  --wandb_group kutta-te-constraint --seed 42 \
  ... (same flags without --kutta_hard_constraint) \
  --kutta_loss_weight 1.0 --kutta_region_k 8

# Variant 3: Hard only — run if time permits
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent edward --wandb_name "edward/kutta-hard-only-s42" \
  --wandb_group kutta-te-constraint --seed 42 \
  ... --kutta_hard_constraint --kutta_region_k 8
```

Run the best variant with seed 73 before submitting for review.

**Note:** The `compute_te_features` max-x approach may find a single TE node if upper/lower TE nodes are coincident. K=8 neighbors naturally cover both TE sides regardless — start with K=8 and report how many distinct nodes you find near the TE to inform K selection.

## Baseline

Current best (PR #2357 — Vortex-Panel Induced Velocity, 2-seed avg):

| Metric | Value | Target |
|--------|-------|--------|
| p_in | 11.872 | < 11.872 |
| p_oodc | 7.459 | < 7.459 |
| **p_tan** | **26.319** | **< 26.319** |
| p_re | 6.229 | < 6.229 |

W&B: aycq1m8m (seed 42), 9sk276v6 (seed 73).

Reproduce baseline:
```bash
cd cfd_tandemfoil && python train.py \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### Variant 1: Hard+Soft (kutta_hard_constraint + kutta_loss_weight=0.5, K=8) — 2-seed

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| p_in | 12.1 | 12.0 | **12.05** | 11.872 | **+1.5%** ❌ |
| p_oodc | 7.2 | 7.5 | **7.35** | 7.459 | **-1.5%** ✅ |
| p_tan | 26.7 | 27.3 | **27.0** | 26.319 | **+2.6%** ❌ |
| p_re | 6.2 | 6.7 | **6.45** | 6.229 | **+3.5%** ❌ |

- W&B: qadoavcn (seed 42, best epoch 154), xu7hox1a (seed 73, best epoch 155)
- Peak memory: 40.4 GB
- Training time: ~180 min (154-155 epochs at 70s/epoch)

### Variant 2: Soft only (kutta_loss_weight=1.0, K=8) — seed 42 only

| Metric | Value | Baseline | Δ |
|--------|-------|----------|---|
| p_in | 12.7 | 11.872 | +7.0% ❌ |
| p_oodc | 7.3 | 7.459 | -2.1% ✅ |
| p_tan | 27.9 | 26.319 | +6.0% ❌ |
| p_re | 6.4 | 6.229 | +2.7% ❌ |

- W&B: qp5l0zuz (seed 42, best epoch 154)

### Variant 3: Hard only (kutta_hard_constraint, no soft loss, K=8) — seed 42 only

| Metric | Value | Baseline | Δ |
|--------|-------|----------|---|
| p_in | 12.2 | 11.872 | +2.8% ❌ |
| p_oodc | 7.3 | 7.459 | -2.1% ✅ |
| p_tan | 27.5 | 26.319 | +4.5% ❌ |
| p_re | 6.1 | 6.229 | -2.1% ✅ |

- W&B: 223s9pb2 (seed 42, best epoch 154)

### Commands used

```bash
# Variant 1: Hard+Soft (best variant, run with both seeds)
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent edward --wandb_name "edward/kutta-hard-soft-s42" \
  --wandb_group kutta-te-constraint --seed 42 \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --kutta_hard_constraint --kutta_loss_weight 0.5 --kutta_region_k 8
```

### What happened

**The Kutta TE constraint did not improve surface pressure metrics.** All 3 variants show p_tan regression (+2.6% to +6.0%) and p_in regression (+1.5% to +7.0%). The only consistent improvement is p_oodc (-1.5% to -2.1%).

Analysis:
1. **The hard constraint (averaging K=8 nearest TE-region pressure predictions) likely over-smooths the TE pressure distribution.** Real airfoil TE pressure gradients are steep — the upper and lower surface pressures converge rapidly but aren't necessarily equal at the discrete mesh nodes near TE. Forcing them to be equal may remove physically meaningful pressure gradients.

2. **The soft loss (max-min spread penalty) similarly penalizes the natural pressure variation near TE.** The Kutta condition requires pressure continuity at the exact TE point, not across K=8 neighboring nodes spanning both upper and lower surfaces.

3. **K=8 may be too large.** With 8 nodes spanning both sides of the TE, we're averaging across the upper-lower pressure jump region. K=2-4 (just the closest nodes to the exact TE) might be more appropriate. Alternatively, the constraint could target only the pressure difference between the single upper and single lower TE node, rather than averaging a neighborhood.

4. **The model may already implicitly learn near-TE pressure continuity** through the existing te_coord_frame features and the strong surface MAE loss signal, making the explicit Kutta constraint redundant and potentially harmful.

### Suggested follow-ups

- Try K=2 or K=4 (much smaller neighborhood, closer to the actual TE point)
- Instead of averaging all K nodes, compute the constraint as the difference between the mean of upper-surface TE nodes and mean of lower-surface TE nodes (true Kutta condition: ΔCp_upper = ΔCp_lower at TE)
- Try much smaller kutta_loss_weight (0.01-0.1) to provide gentle regularization without overriding the main surface MAE signal
- Consider applying the constraint only to tandem foils where p_tan is the weakest metric

**Comment by tcapelle** [comment]
ADVISOR: Good analysis, Edward. The p_oodc improvement (-1.5 to -2.1%) across all 3 variants confirms the Kutta constraint provides useful physics information. But K=8 is over-smoothing — you're averaging across the upper/lower pressure jump region near TE, which destroys physically meaningful gradients.

## v2 Instructions: Refined Kutta Constraint

The core issue is that K=8 captures nodes on BOTH sides of the TE, averaging across the pressure discontinuity. The true Kutta condition is: pressure on the upper surface TE = pressure on the lower surface TE at a single point.

### Changes for v2:

1. **Reduce K from 8 to 2** (just the single closest node on each side of the TE). If your implementation uses KNN to find K nearest nodes to the TE, K=2 should grab one upper and one lower node.

2. **Reformulate the hard constraint**: Instead of averaging all K nodes, compute the true Kutta condition as the absolute difference between the upper-surface TE node prediction and the lower-surface TE node prediction, then average these two predictions as the constrained TE value. This preserves the correct physics.

3. **Lower the soft loss weight**: Try `--kutta_loss_weight 0.05` (was 0.5). The p_oodc gain is already strong at high weight — at lower weight, the constraint should provide gentler regularization without overriding the main loss on p_in/p_tan/p_re.

4. **Apply to tandem foils only**: Since p_tan is our hardest metric (26.3 baseline) and tandem TE effects are strongest due to foil-foil interference, add a flag to only apply the Kutta constraint to the tandem (rear) foil. This reduces the 'damage' to p_in while targeting p_tan improvement.

### Configs to try (use `--wandb_group kutta-v2`):

**Config A** (refined K=2, low weight, tandem-only):
```bash
python train.py --agent edward --wandb_name 'edward/kutta-v2-cfgA-s42' \
  --wandb_group kutta-v2 --seed 42 \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --kutta_hard_constraint --kutta_loss_weight 0.05 --kutta_region_k 2 --kutta_tandem_only
```

**Config B** (K=2, higher weight, tandem-only):
Same as A but `--kutta_loss_weight 0.2 --wandb_name 'edward/kutta-v2-cfgB-s42'`

**Config C** (K=2, low weight, both foils):
Same as A but remove `--kutta_tandem_only`, `--wandb_name 'edward/kutta-v2-cfgC-s42'`

Run all 3 configs with seed 42 first. If any beat baseline on 2+ metrics, run seed 73.

Note: you may need to implement `--kutta_tandem_only` and `--kutta_region_k 2` modifications. The key change is in how TE nodes are selected — use K=2 so you get exactly 1 upper + 1 lower node, and reformulate the averaging.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2374 because ADVISOR: Closing — Edward has been reassigned. The original Kutta v1 showed p_oodc -1.5%, but v2 results were never posted. Edward will receive a new bold Round 48 experiment assignment..
